server/src/main/java/org/apache/cloudstack/diagnostics/DiagnosticsServiceImpl.java

```diff
@@ -126,10 +185,336 @@ protected String prepareShellCmd(String cmdType, String ipAddress, String option
+        }
+        cmd.setAccessDetail(accessDetails);
+    }
+
+    private Answer prepareDiagnosticsFilesInSystemVm(VMInstanceVO vmInstance, List<String> fileList) {
+        final PrepareFilesCommand cmd = new PrepareFilesCommand(fileList, DataRetrievalTimeout.value());
+        configureNetworkElementCommand(cmd, vmInstance);
+        Answer answer = agentManager.easySend(vmInstance.getHostId(), cmd);
+        return answer;
+    }
+
+    private Answer deleteDiagnosticsZipFileInsystemVm(VMInstanceVO vmInstance, String zipFileName) {
+        final DeleteFileInVrCommand cmd = new DeleteFileInVrCommand(zipFileName);
+        configureNetworkElementCommand(cmd, vmInstance);
+        final Answer fileCleanupAnswer = agentManager.easySend(vmInstance.getHostId(), cmd);
+        if (fileCleanupAnswer == null) {
+            LOGGER.error(String.format("Failed to cleanup diagnostics zip file on vm: %s", vmInstance.getUuid()));
+        } else {
+            if (!fileCleanupAnswer.getResult()) {
+                LOGGER.error(String.format("Zip file cleanup for vm %s has failed with: %s", vmInstance.getUuid(), fileCleanupAnswer.getDetails()));
+            }
+        }
+
+        return fileCleanupAnswer;
+    }
+
+    /**
+     * Generate a list of diagnostics file to be retrieved depending on the system VM type
+     *
+     * @param optionalFileList Optional list of files that user may want to retrieve, empty by default
+     * @param vmInstance       system VM instance, either SSVM, CPVM or VR
+     * @return a list of files to be retrieved for system VM, either generated from defaults depending on the VM type, or specified
+     * by the optional list param
+     */
+    private List<String> getFileListToBeRetrieved(List<String> optionalFileList, VMInstanceVO vmInstance) {
+        DiagnosticsFilesList fileListObject = DiagnosticsFilesListFactory.getDiagnosticsFilesList(optionalFileList, vmInstance);
+        List<String> fileList = new ArrayList<>();
+
+        if (fileListObject != null) {
+            fileList = fileListObject.generateFileList();
+        }
+        return fileList;
+    }
+
+    private Pair<Boolean, String> orchestrateCopyToSecondaryStorageNonVMware(final DataStore store, final String vmControlIp, String fileToCopy, Long vmHostId) {
+        CopyToSecondaryStorageCommand toSecondaryStorageCommand = new CopyToSecondaryStorageCommand(store.getUri(), vmControlIp, fileToCopy);
+        Answer copyToSecondaryAnswer = agentManager.easySend(vmHostId, toSecondaryStorageCommand);
+        Pair<Boolean, String> copyAnswer;
+        if (copyToSecondaryAnswer != null) {
+            copyAnswer = new Pair<>(copyToSecondaryAnswer.getResult(), copyToSecondaryAnswer.getDetails());
+        } else {
+            copyAnswer = new Pair<>(false, "Diagnostics Zip file to secondary storage failed");
+        }
+        return copyAnswer;
+    }
```

Comment:
```orchestrateCopyToSecondaryStorageNonVMware``` is a misleading name for this method as it actually does the provisioning and not (just) the orchestration.
The `orchestrate` method would just decide on things, the `provision`ing would apply it remotely
Can you change/explain?

Reply:
Changing.